### ISSUE TITLE
[Feature Request]: Structured Output

### ISSUE THREAD

**Issue by Hk669** [ag2ai/ag2#23]
### Is your feature request related to a problem? Please describe.

Ive came across many comments from the community, asking for the formatted output/ structural output from the Agentic chat (initate_chat and groupchat ). I believe generating the expected response after the agentic chat can be the biggest improvement in AutoGen. 

Me, @marklysze and @sonichi discussed about adding this feature over a call. 

### Describe the solution you'd like

There should be an output message after the completion of the chat, i can see there is `last_message`, but that doesn't serve the exact feature. adding this feature can help production apps use AutoGen.

solution: TBD ( help and ideas are appreciated )

### Additional context

We can get started with the openai structured output and final goal should be the structured output for the complete chat.

cc @sonichi @BeibinLi @qingyun-wu @marklysze @yiranwu0 @skzhang1

**Comment by ohdearquant** [comment]
TLDR
- use a custom pydantic model to do field management
- assemble new model on fly via pydantic `create_model`


---
(claude version)
# FieldModel Approach for Structured Outputs

## Core Concept

The proposed approach centers around `FieldModel`, a specialized BaseModel that encapsulates field information and behavior. Unlike traditional FieldInfo objects, FieldModel serves as a more comprehensive container that carries both structure and behavior through the system.

## Key Architecture Components

The architecture consists of three main layers working together to provide flexible yet structured output handling:

### Field Model Layer

FieldModel instances serve as self-contained units that capture all aspects of a field:

```python
class FieldModel(BaseModel):
    name: str
    description: str
    prompt: str
    validator: Callable
    # ... other field metadata
```

Each FieldModel maintains its own prompt and validation logic, allowing fields to be treated as independent, reusable components that carry their behavior with them throughout the system.

### Dynamic Model Generation 

The system leverages Pydantic's `create_model` function to generate models at runtime based on collected FieldModels. This dynamic generation enables flexible model composition while maintaining type safety and validation.

### Schema Management

The system includes comprehensive schema handling:

1. Recursive annotation processing for nested models
2. JSON schema generation for LLM context
3. Template integration for instruction generation

## Operational Flow

1. FieldModels are created and configured with prompts and validators
2. Runtime models are generated from FieldModel collections
3. Schema information is extracted and injected into instruction templates
4. LLM inference is performed using generated models
5. Results are validated using field-specific logic

## Benefits

- **Modular**: Fields are self-contained units with their own behavior
- **Flexible**: Runtime model generation enables dynamic composition
- **Maintainable**: Validation and prompts stay with their respective fields
- **Extensible**: New field types can be easily added
- **Type-Safe**: Leverages Pydantic's type system

This approach provides a powerful foundation for structured outputs while maintaining flexibility and extensibility in the system.

**Comment by marklysze** [comment]
Thanks for raising the PR and your comments in here, @ohdearquant - let us know your thoughts on how that could be integrated into AG2 (e.g. is it specified as part of the [initiate_chat summary parameters](https://ag2ai.github.io/ag2/docs/reference/agentchat/conversable_agent#initiate_chat))?

Let's discuss this feature here and then implement within the PR (so we don't have discussions in two places).

**Comment by ohdearquant** [comment]
no I was more thinking around adding the structure to the agent set up. For example, you might want engineer to always return in certain structure, scientist to return in another structure

**Comment by marklysze** [comment]
Thanks for that @ohdearquant ...

What are your thoughts around:
- Does the structured output have to be json?
- Does the agent have to always return in that structure?
- On the receiving side would we use the similar thing to have agents that expect a structured input?
- Is that structure ultimately generated from the LLM's response, a tool call, or a function?
- What would a basic program look like?

Sorry if you've answered a few of these in your FieldModel comment above.

**Comment by ohdearquant** [comment]
1. json / xml should work similar, but the pydantic structure is easier with json, especially with fuzzy parsing and fuzzy validating fields, 
2. no it does not, we can instruct it to use the format guidance or not, 
3. that will be use case dependent, a lot of usage of the structured output is to interact with external environment, or if you want specific handling of the output, such as giving commands to different agents with different prompts, a structured output would be great here
4. from LLM's response. using chat completion is enough, since they are from LLM response, they can be flexible, we can even have multiple pydantic model from one response
5. please check [here](https://github.com/lion-agi/lion-os/blob/53334910720856c134f04747b95e905f6eb65dc0/notebooks/plan.ipynb) for an example of how output will look like and can be used, and check [here](https://github.com/lion-agi/lion-os/blob/53334910720856c134f04747b95e905f6eb65dc0/lion/operations/brainstorm/brainstorm.py) for a sample algorithm on how to interact with such structured output. and you can check the prompt template logic [here](https://github.com/lion-agi/lion-os/blob/53334910720856c134f04747b95e905f6eb65dc0/lion/core/communication/instruction.py), also check, [FieldModel ](https://github.com/lion-agi/lion-os/blob/53334910720856c134f04747b95e905f6eb65dc0/lion/core/models.py).

No worries,

**Comment by ohdearquant** [comment]
oops

**Comment by marklysze** [comment]
Thanks @ohdearquant!

@Hk669 can you provide some of your thoughts on this as well.

**Comment by ohdearquant** [comment]
moved all the core logic for structured output to the PR now, will start assembling them with existing AG components.
